torch/_lobpcg.py

```diff
@@ -11,7 +11,338 @@
 from .overrides import has_torch_function, handle_torch_function
 
 
-__all__ = ['lobpcg']
+__all__ = ['lobpcg', 'lobpcg2']
+
+def _symeig_backward_complete_eigenspace(D_grad, U_grad, A, D, U):
+    # compute F, such that F_ij = (d_j - d_i)^{-1} for i != j, F_ii = 0
+    F = D.unsqueeze(-2) - D.unsqueeze(-1)
+    F.diagonal().fill_(float('inf'))
+    F.pow_(-1)
+
+    # A.grad = U (D.grad + (U^T U.grad * F)) U^T
+    Ut = U.transpose(-1, -2).contiguous()
+    res = torch.matmul(
+        U,
+        torch.matmul(
+            torch.diag_embed(D_grad) + torch.matmul(Ut, U_grad) * F,
+            Ut
+        )
+    )
+
+    return res
+
+
+def _polynomial_coefficients_given_roots(roots):
+    """
+    Given `roots` of a polynomial, find the polynomial's coefficients.
+
+    If roots = (r_1, ..., r_n), then the method returns
+    coefficients (a_0, a_1, ..., a_n (== 1)) so that
+    p(x) = (x - r_1) * ... * (x - r_n)
+         = x^n + a_{n-1} * x^{n-1} + ... a_1 * x_1 + a_0
+
+    Note: for better performance requires writing a low-level kernel
+    """
+
+    poly_order = roots.shape[-1]
+    poly_coeffs_shape = list(roots.shape)
+    # we assume p(x) = x^n + a_{n-1} * x^{n-1} + ... + a_1 * x + a_0,
+    # so poly_coeffs = {a_0, ..., a_n, a_{n+1}(== 1)}
+    poly_coeffs_shape[-1] += 1
+    poly_coeffs = roots.new_zeros(poly_coeffs_shape)
+    poly_coeffs[..., -1] = 1
+
+    # perform the Horner's rule
+    for i in range(1, poly_order + 1):
+        for j in range(poly_order - i - 1, poly_order):
+            poly_coeffs[..., j] -= roots[..., i - 1] * poly_coeffs[..., j + 1]
```

Comment:
When I try using `gradgradcheck` with `anomaly_detection=True`, it points to these lines complaining about the in-place operations. @alban, could you know the reason?

Reply:
Doing `foo[j] = bar` or `foo -= bar` when foo is a Tensor are both inplace operations.

The problem is that the multiplication on that line needs the value of `poly_coeffs` to compute the backward but you change it inplace.
Here is a version that will work (at the cost of an extra clone when we require gradients) and with the inner loop removed:
```python
import torch

roots = torch.rand(2, 4)
print(roots)

def original(roots):
    poly_order = roots.shape[-1]
    poly_coeffs_shape = list(roots.shape)
    # we assume p(x) = x^n + a_{n-1} * x^{n-1} + ... + a_1 * x + a_0,
    # so poly_coeffs = {a_0, ..., a_n, a_{n+1}(== 1)}
    poly_coeffs_shape[-1] += 1
    poly_coeffs = roots.new_zeros(poly_coeffs_shape)
    poly_coeffs[..., -1] = 1

    # perform the Horner's rule
    for i in range(1, poly_order + 1):
        for j in range(poly_order - i - 1, poly_order):
            poly_coeffs[..., j] -= roots[..., i - 1] * poly_coeffs[..., j + 1]

    return poly_coeffs

original_res = original(roots)


# Fast implementation
def fast(roots):
    poly_order = roots.shape[-1]
    poly_coeffs_shape = list(roots.shape)
    # we assume p(x) = x^n + a_{n-1} * x^{n-1} + ... + a_1 * x + a_0,
    # so poly_coeffs = {a_0, ..., a_n, a_{n+1}(== 1)}
    poly_coeffs_shape[-1] += 2
    poly_coeffs = roots.new_zeros(poly_coeffs_shape)
    poly_coeffs[..., 0] = 1
    poly_coeffs[..., -1] = 1

    # perform the Horner's rule
    for i in range(1, poly_order + 1):
        new = poly_coeffs.clone() if roots.requires_grad else poly_coeffs
        out = new.narrow(-1, poly_order - i, i + 1)
        out -= roots.narrow(-1, i - 1, 1) * poly_coeffs.narrow(-1, poly_order - i + 1, i + 1)
        poly_coeffs = new

    return poly_coeffs.narrow(-1, 1, poly_order + 1)

fast_res = fast(roots)

print(original_res)
print(fast_res)
print((original_res - fast_res).abs().sum())

roots.requires_grad_(True)
fast(roots).sum().backward()
print("Backward ran fine")
```


Also nit. doing `poly_coeffs.select(-1, idx)` will most likely be a bit faster than `poly_coeffs[..., idx]` as it avoids advance indexing logic and parsing.

Reply:
Thank you! Yeah, the `roots` do require grads, so, I guess, this is what triggers the warning.

Reply:
@albanD , the detached `roots` solved the issue. While `gradcheck` passes, `gradgradcheck` fails.
Does `gradgradcheck` check the gradients for `backward`?

Reply:
>  the detached roots

Does that mean you `.detach()` the roots? If so that sounds wrong and would explain the gradgradcheck issue.
If not, this is surprising. A "simple hack" you can use here is to run gradcheck  on `_symeig_backward`. That should give you an example that will fail when gradgradcheck of lobpcg fails and that is much simpler to debug as there is a single backward.

Reply:
Well, I pass a detached roots to the function. In fact, the polynomial coefficients are indeed a function of its roots, but why should it matter in forward?

Reply:
If you detach the roots, then that would explain why gradcheck works fine. Because .detach() does not change the value computed by the backward pass.
But gradgradcheck fails because you did not create the proper graph during the first backward due to the .detach().

Reply:
Super weird! When I use your method I am able to pass the `gradgradcheck` with some tweaking of `eps`.
In your code you use `clone` in the loop. Could you please explain it? Is it efficient?

Reply:
`.clone()` is differentiable, `.detach()` is not. So using `.clone()` won't cause any gradient issue while using `.detach()` will explicitly break the graph.
Note that *only* .detach() (or code in torch.no_grad() block) break the graph, every other (differentiable) function will either compute gradients or return an error when not implemented.

`.clone()` will do a memory copy so it is not great not but I can't think of a way to make this look work well :/ 
The only option if you don't want the clone is to write this `_polynomial_coefficients_given_roots` as a custom function (that can do what it wants in terms of inplace in the forward) and write a custom backward for it haha. But you might need some of the intermediary results to write the backward :/

Reply:
OK, I expected that `forward` is just a single node in the computational graph, but, apparently, this example with `roots` (I use `poly = _poly(D)` in forward where `D.requires_grad == True`) shows this is not the case. Hmm, I am super confused. From the documentation on forward:

forward() - the code that performs the operation. It can take as many arguments as you want, with some of them being optional, if you specify the default values. All kinds of Python objects are accepted here. Tensor arguments that track history (i.e., with requires_grad=True) will be converted to ones that don’t track history before the call, and their use will be registered in the graph.

So, I guess, I have to understand the difference between `history` and `registration in the graph`.

Reply:
Yeah, the custom `poly_from_roots` is out, because its beckward will require finding the roots of a polynomial ;)

Reply:
> I expected that forward is just a single node in the computational graph

It is.
But here we are doing double backward:
So the `forward` you implement is a single node. And it's backward is running your `backward` function. And since this backward is not marked `once_differentiable`, for the double backward, we just run the `backward` with grad mode enabled and use the autograd to get the double backward.

So `_polynomial_coefficients_given_roots` that is used in your `backward` runs during the first backward pass. But is evaluated during the "forward" part of the double backward pass.
If it was used in the `forward` of the custom function, you would not have this problem indeed!